vita3k/main.cpp

```diff
@@ -86,16 +86,70 @@ static void run_execv(char *argv[], EmuEnvState &emuenv) {
 int main(int argc, char *argv[]) {
     ZoneScoped; // Tracy - Track main function scope
     Root root_paths;
-    root_paths.set_base_path(string_utils::utf_to_wide(SDL_GetBasePath()));
-    root_paths.set_pref_path(string_utils::utf_to_wide(SDL_GetPrefPath(org_name, app_name)));
+    auto base_path = SDL_GetBasePath();
+    auto pref_path = SDL_GetPrefPath(org_name, app_name);
+    root_paths.set_base_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_pref_path(string_utils::utf_to_wide(pref_path));
+    root_paths.set_config_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_shared_path(string_utils::utf_to_wide(base_path));
+    root_paths.set_cache_path(string_utils::utf_to_wide(base_path));
+    SDL_free(base_path);
+    SDL_free(pref_path);
+
+#if defined(__linux__) && !defined(__ANDROID__) && !defined(__APPLE__)
+    // XDG Data Dirs.
+    auto env_home_dir = getenv("HOME");
+    auto env_data = getenv("XDG_DATA_HOME");
+    if (env_data != NULL)
+        root_paths.set_pref_path(string_utils::utf_to_wide(env_data) / fs::path(app_name) / fs::path(app_name));
+
+    env_data = getenv("XDG_CONFIG_HOME");
+    if (env_data != NULL)
+        root_paths.set_config_path(string_utils::utf_to_wide(env_data) / fs::path(app_name));
+    else if (env_home_dir != NULL)
+        root_paths.set_config_path(fs::path(string_utils::utf_to_wide(env_home_dir)) / ".config" / fs::path(app_name));
+
+    env_data = getenv("XDG_CACHE_HOME");
+    if (env_data != NULL)
+        root_paths.set_cache_path(string_utils::utf_to_wide(env_data) / fs::path(app_name));
+    else if (env_home_dir != NULL)
+        root_paths.set_cache_path(fs::path(string_utils::utf_to_wide(env_home_dir)) / ".cache" / fs::path(app_name));
+
+    if (env_home_dir != NULL) {
+        // Don't assume that base_path is portable.
+        root_paths.set_shared_path(fs::path(env_home_dir) / ".local/share" / fs::path(app_name));
+    }
+    env_data = getenv("XDG_DATA_DIRS");
+    if (env_data != NULL) {
+        auto env_paths = string_utils::split_string(env_data, ':');
+        for (std::vector<std::string>::iterator iter = env_paths.begin() ; iter != env_paths.end() ; iter++) {
+            if (fs::exists(*iter / fs::path(app_name))) {
+                root_paths.set_shared_path(string_utils::utf_to_wide(*iter) / fs::path(app_name));
+                break;
+            }
+        }
+    }
+#endif
+
+    // Create default preference and cache path for safety
+    if (!fs::exists(root_paths.get_config_path()))
+        fs::create_directories(root_paths.get_config_path());
```

Comment:
It would be nice for all of this to go inside its own function (I think you can put it in app_init.cpp) which would get called here.